### DISCUSSION TITLE
Loratap Garage Door Opener

### DISCUSSION THREAD

**Discussion by illuzn** [make-all/tuya-local#1419] [Ideas]
This is a very simple garage door opener with one DPS that operates as a push button which open/stop/closes the garage (I have an old garage door opener which has one button which functions as open/stop/close).

Being so simple, there is no way for Home Assistant to know if it is opening/ closing or if it is somehow stuck half way between open and closed.

This causes a potential problem for me in the following scenario:
1. Garage door is open.
2. User pushes close.
3. There is an obstruction in the way (maximum force exceeded) , the infrared sensor which detects something in the way is closed, or somebody uses a physical remote or the push button to stop closing.
4. Result: The garage door is stuck in the closing state permanently and unable to be opened or closed.

The opposite can occur when opening the garage but it is more of an edge case because the first 2 exceptions in item 3 generally won't apply (maximum force will only be exceeded if someone is hanging onto the garage door and the infrared sensors do nothing on garage open).

I have two possible solutions for this.
1. Quick and dirty. Map the open/close dps also to a stop dps. To issue the stop command all it needs to do is flip the current result of dps 1 i.e. if dps 1 is currently `true` set to `false` to stop and vice versa. This will allow pushing the stop button to function as a stop button as well as trigger the garage in case it is stopped.
2. Use the Travel Time configured in dps 102.  This is obviously more complicated and involves setting up a timer.  Here is some pseudocode for how I think that would work:

```
On Open/Close button pushed:
   Destroy any previous Asynchronous Timers (if any).
   Create Asynchronous Timer and tell it Travel Time in dps 102, whether open/ close was pushed.

Asynchronous timer:
  If close was pushed:
     Wait Travel Time in dps 102 seconds.
     If open in dps 101 is closed:
         Clear problem binary sensor.
      If open in dps 101 is open:
         Set cover state to open. # This will permit pressing close again.
         Set a new problem binary sensor
   If open was pushed:
      Wait Travel Time in dps 102 seconds.
      If open in dps 101 is open:
         Clear problem binary sensor.
      If open in dps 101 is closed:
         Set cover state to closed. # This will permit pressing close again.
         Set a new problem binary sensor.
```

In particular, the warning flag will be helpful because you can send a notification to the user to say "error opening/ closing garage, please make sure it's open/closed". This is the behavior in the app currently and is quite useful.

**Comment by make-all** [comment]
From the description, I think you can only expose a button, as there is insufficient info from the device to implement a cover entity
